fw/tls.c

```diff
@@ -797,6 +797,7 @@ tfw_tls_conn_send(TfwConn *c, TfwMsg *msg)
 {
 	int r;
 	TlsCtx *tls = tfw_tls_context(c);
+	int ss_flags = msg->ss_flags;
```

Comment:
Usually I find your commits commented pretty well. In this case, though, I would really appreciate a more comprehensive commit message (or a block comment here in the code) as it's not really clear (to me) exactly what, why and how happens. Perhaps, it's just me though. Thanks.

Reply:
The commit already has a blocked message besides the first line, which describes why the flag should be saved in the first place. Also, the issue reference link will have detailed background information.

Reply:
Yeah, that's exactly what I was talking about. I didn't understand the commit message which should be self-sufficient and clear enough. From what I can see, we are in the sending function, and a specific message is given to it for sending. Which means no one else can work on it, doesn't it? How would accessing the flags just a few cycles earlier (saving the flags) significantly change the situation? I admit that I don't know this code well enough, but I believe that the commit message should be more comprehensive than it is now. Thanks.

Reply:
Sending to the client (downstream) is asynchronous, and `tfw_tls_conn_send` is in the path of receiving the response from the backend (upstream), and does not actually send.

The CPU is not always exclusively occupied by softirqs, and other hardware interrupts may also preempt the CPU with an indeterminate period. On the other hand, sending (`ss_tx_action`) is not necessarily triggered by an IPI or even softirq, and can be processed quickly if the softirq handler is already running. So, in summary, there is likely a small window that allows the message to be sent and released faster than the next `if` statement in `tfw_tls_conn_send`.

---

Regarding the clarification of the commit message, I think we cannot or are not suitable to describe everything in the commit message unless the issue is extremely simple. Even 10,000 words cannot make it clear to readers without any prior background knowledge. However, we do have to define the commit message to make it clear what it does or why it does it. And I think if the reviewer really does a serious review, he should at least read the issue comments first, or even read some related code to get a prior understanding. And, to be honest, I have no idea how to convert a long issue comment page into a commit message or even code comments: https://github.com/tempesta-tech/tempesta/issues/2169#issuecomment-2247592874.

Last but not least, no offense, please look at the bugfix first and think about the correctness, which is more important than the commit message style.

Reply:
I can't understand how this  fix help. If msg->ss_flags can be changed on other CPU it can occurs also at line msg->ss_flags |= SS_SKB_TYPE2F(TTLS_MSG_APPLICATION_DATA) or when it is passed to `ss_send`. Moreover there is a chance that it can be canged just before you save it in a local variable. At what place  (what function) changes `ss_flags` on other CPU?

Reply:
@EvgeniiMekhanik -- Yeah, thanks. I failed in making this clear enough.

Reply:
@EvgeniiMekhanik in my understanding the `msg` is deleted by a callback form the TCP transmission, which may happen on another CPU thanks to scheduling logic in `ss_send()`. Doesn't https://github.com/tempesta-tech/tempesta/issues/2169#issuecomment-2247592874 answer your question?